upgrade.c

```diff
@@ -0,0 +1,401 @@
+}
+
+static int install_program_fragment(fota_candidate_iterate_callback_info *info)
+{
+    int ret;
+    static const uint32_t  print_range_percent = 5;
+
+    uint32_t progress = (info->frag_pos + info->frag_size) * 100 / info->header_info->fw_size;
+    uint32_t prev_progress = info->frag_pos * 100 / info->header_info->fw_size;
+
+    if (info->frag_pos == 0 || ((progress / print_range_percent) > (prev_progress / print_range_percent))) {
+        pr_info("Flashing. %" PRIu32 "%c complete", progress, '%');
+    }
+
+    // Would need to check that info->frag_buf is aligned to 8 bytes, as HAL flash driver won't accept it
+    // otherwise. We have preliminary knowledge that it's like that here, so we slip this check.
+    ret = flash_program_page(&flash_obj, MBED_CONF_MBED_BOOTLOADER_APPLICATION_START_ADDRESS + info->frag_pos,
+                             info->frag_buf, info->frag_size);
+    if (ret) {
+        pr_error("flash program failed");
+        return FOTA_STATUS_STORAGE_WRITE_FAILED;
+    }
+    return FOTA_STATUS_SUCCESS;
+}
+
+static int install_finish(const fota_candidate_iterate_callback_info *info)
+{
+    int ret;
+    uint32_t header_buf_size = fota_align_up(INTERNAL_HEADER_SIZE, flash_page_size);
+    uint8_t *header_buf = (uint8_t *) malloc(header_buf_size);
+    if (!header_buf) {
+        pr_error("Unable to allocate header");
+        return FOTA_STATUS_OUT_OF_MEMORY;
+    }
+    fota_header_info_t *header = (fota_header_info_t *) header_buf;
+
+    memset(header_buf, 0, header_buf_size);
+    memcpy(header, info->header_info, INTERNAL_HEADER_SIZE);
+
+    // Flash new FW header
+    ret = flash_program_page(&flash_obj, MBED_CONF_MBED_BOOTLOADER_APPLICATION_HEADER_ADDRESS,
+                             header_buf, header_buf_size);
+    if (ret) {
+        pr_error("Programming header failed");
+        return FOTA_STATUS_STORAGE_WRITE_FAILED;
+    }
+
+    return FOTA_STATUS_SUCCESS;
+}
+
+static int install_iterate_handler(fota_candidate_iterate_callback_info *info)
+{
+
+    switch (info->status) {
+        case FOTA_CANDIDATE_ITERATE_START:
+            return install_start(info);
+        case FOTA_CANDIDATE_ITERATE_FRAGMENT:
+            return install_program_fragment(info);
+            break;
+        case FOTA_CANDIDATE_ITERATE_FINISH:
+            return install_finish(info);
+            break;
+        default:
+            return FOTA_STATUS_INTERNAL_ERROR;
+    }
+    return 0;
+}
+
+static int check_and_install_update()
+{
+    int ret = FOTA_STATUS_NOT_FOUND;
+    bool validate = true;
+    bool force_encrypt = false;
+#if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
+    uint8_t fw_key[FOTA_ENCRYPT_KEY_SIZE];
+    ret = fota_nvm_fw_encryption_key_get(fw_key);
+    if (ret) {
+        pr_info("Candidate not found");
+        return FOTA_STATUS_NOT_FOUND;
+    }
+    memset(fw_key, 0, sizeof(fw_key));
+    force_encrypt = true;
+#endif // #if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
```

Comment:
i think it is "safer" to initialize `force_encrypt` value at declaration.
```suggestion

#if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
    bool force_encrypt = true;
    uint8_t fw_key[FOTA_ENCRYPT_KEY_SIZE];
    ret = fota_nvm_fw_encryption_key_get(fw_key);
    if (ret) {
        pr_info("Candidate not found");
        return FOTA_STATUS_NOT_FOUND;
    }
    memset(fw_key, 0, sizeof(fw_key));
    force_encrypt = true;
#else
    bool force_encrypt = false;
#endif // #if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
```

Reply:
I'll change, but remove line 215
why it's safer?

Reply:
not sure :) FI mitigations

Reply:
removed line 215